### ISSUE TITLE
A major bug in the inference code

### ISSUE THREAD

**Issue by prilgrim-whtie** [LiAutoAD/LightVLA#26]
Hello, thank you for your tremendous contributions to this field. However, while reviewing the code, we discovered a significant bug that may lead to incorrect results in the paper. In the inference code, I noticed the line `actions_hidden_states = last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]`. This is unreasonable, as the selected `actions_hidden_states` do not match the `actions_hidden_states` used during training.
https://github.com/LiAutoAD/LightVLA/blob/a4680fda5ffe73029190ac97328aa34b0e87a45a/prismatic/extern/hf/modeling_prismatic.py#L1241-L1245

During training:
1. First, take the `ground_truth_token_ids` from the dataset, then use the two functions `current_action_mask=get_current_action_mask(ground_truth_token_ids)` and
`next_actions_mask=get_next_actions_mask(ground_truth_token_ids)` to obtain `actions_mask([current_action_mask  | next_actions_mask])` (note that `actions_mask` includes the stop token, i.e., `[......, True,True,True,True,True,True,True,False]) 
https://github.com/LiAutoAD/LightVLA/blob/a4680fda5ffe73029190ac97328aa34b0e87a45a/vla-scripts/finetune.py#L344-L346
 
2. For the model’s output, first remove the stop token, then use `actions_mask` to obtain `actions_hidden_states`. In inference, this should correspond to: `actions_hidden_states = last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+2):-2, :,], rather than actions_hidden_states=last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,].
https://github.com/LiAutoAD/LightVLA/blob/a4680fda5ffe73029190ac97328aa34b0e87a45a/vla-scripts/finetune.py#L376-L385

When I changed `last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]` to `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+2):-2, :,]` in the inference code, I noticed that the model's performance declined.

**Comment by Promisery** [comment]
Thank you for your insightful observation. However, since stop token is only a single token, `last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]` should be changed to `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-1, :,]`. Your solution, by using `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+2):-2, :,]`, is using the last text token, which leads to performance degradation. In our experiments, we find negligible performance discrepancy after changing `last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]` to `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-1, :,]`. Therefore, this bug does little damage the performance of LightVLA.

**Comment by Promisery** [comment]
A bug should theoretically damage the performance of the model (unless it is a data leakage bug, which is impossible for LIBERO since the tests are conducted in simulation environments). Fixing the bug should improve (or at least not damage) the performance of the model. Therefore, this bug can only positively affect the conclusions of our paper - it proves the robustness of LightVLA, as it is able to function normally even with a bug in the inference code.

**Comment by prilgrim-whtie** [comment]
> Thank you for your insightful observation. However, since stop token is only a single token, `last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]` should be changed to `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-1, :,]`. Your solution, by using `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+2):-2, :,]`, is using the last text token, which leads to performance degradation. In our experiments, we find negligible performance discrepancy after changing `last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]` to `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-1, :,]`. Therefore, this bug does little damage the performance of LightVLA.感谢您的深刻见解。然而，由于停止标记只是一个单一标记， `last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]` 应该更改为 `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-1, :,]` 。您的解决方案，通过使用 `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+2):-2, :,]` ，是使用了最后一个文本标记，这会导致性能下降。在我们的实验中，我们发现将 `last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]` 更改为 `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-1, :,]` 后性能差异可以忽略不计。因此，这个错误对 LightVLA 的性能影响很小。

That's not the case. Because openvla-oft is autoregressive, it uses the last token of the prompt and the previous n-1 action tokens to predict the next n action tokens. This is the official approach. Therefore, `last_hidden_states[ :, -ACTION_DIM * NUM_ACTIONS_CHUNK:, :,]` should be changed to `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-2, :,]`, rather than `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-1, :,] `，To verify your point, I tested using `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK+1):-1, :,]` and found that performance also degrades, and the impact on performance is not negligible.

**Comment by prilgrim-whtie** [comment]
> A bug should theoretically damage the performance of the model (unless it is a data leakage bug, which is impossible for LIBERO since the tests are conducted in simulation environments). Fixing the bug should improve (or at least not damage) the performance of the model. Therefore, this bug can only positively affect the conclusions of our paper - it proves the robustness of LightVLA, as it is able to function normally even with a bug in the inference code.理论上，一个错误应该会损害模型的性能（除非它是一个数据泄漏错误，这对 LIBERO 来说是不可能的，因为测试是在模拟环境中进行的）。修复这个错误应该会提高（至少不会损害）模型的性能。因此，这个错误只能对我们的论文结论产生积极影响——它证明了 LightVLA 的鲁棒性，因为它即使在推理代码中存在错误也能正常运行。

In theory, that is correct, but the actual experimental results are exactly the opposite of the theory. Using either `last_hidden_states[ :, -(ACTION_DIM *NUM_ACTIONS_CHUNK+1):-2, :,]` or `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK + 1):-1, :,]` both result in a performance drop, and it is not negligible. It is not as you stated: “This error can only have a positive impact on the conclusions of our paper—it proves the robustness of LightVLA,”

**Comment by prilgrim-whtie** [comment]
If possible, please mark this issue as "active," as I think my issue has not been resolved or verified.Thank  you.

**Comment by Promisery** [comment]
1. OpenVLA-OFT is not autoregressive. It's called parallel decoding. Please read the parallel decoding section of the OpenVLA-OFT paper to better understand how the action generation process of OpenVLA-OFT works. Alternatively, you can inspect the input_embedding of the model and see where the empty (i.e. all zero) embeddings are.
2. In our experiments, LightVLA with `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK + 1):-1, :,]` obtains 97% overall accuracy.
3. Please share your test results (and logs if possible) with both `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK + 1):-1, :,]` and `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK):, :,]`. If you find that LightVLA performs good on several tasks and poorly on others, chances are that there are compatibility issues in the simulation environments, which is beyond our capability to fix.

**Comment by Promisery** [comment]
> In theory, that is correct, but the actual experimental results are exactly the opposite of the theory. Using either `last_hidden_states[ :, -(ACTION_DIM *NUM_ACTIONS_CHUNK+1):-2, :,]` or `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK + 1):-1, :,]` both result in a performance drop, and it is not negligible. It is not as you stated: “This error can only have a positive impact on the conclusions of our paper—it proves the robustness of LightVLA,”

Either `last_hidden_states[ :, -(ACTION_DIM *NUM_ACTIONS_CHUNK+1):-2, :,]` or `last_hidden_states[ :, -(ACTION_DIM * NUM_ACTIONS_CHUNK + 1):-1, :,]` must be correct, and the other one must be wrong. If you are seeing similar results with both configurations, it is likely that the issue lies within your experiments.
